drop Function, rename FunctionBody -> Function

stepancheg opened the pull request:
If someone needs Function to be refcounted, Function can be explicitly
wrapped in sharedpointer, like this:

```
var functionPtr = new(Function((x: Int) -> x * 2));
var thirtyFour = functionPtr^(17);
```

Note that sharedpointers are not thread-safe (thus Function weren't
thread-safe until this patch), so simple function copying could
cause memory problems.

jckarter commented:
Sounds good. Separating the ownership policy from `Function` also allows you to use it with unique pointers, which is useful.

stepancheg committed:
drop Function, rename FunctionBody -> Function

If someone needs Function to be refcounted, Function can be explicitly
wrapped in sharedpointer, like this:

```
var functionPtr = new(Function((x: Int) -> x * 2));
var thirtyFour = functionPtr^(17);
```

Note that sharedpointers are not thread-safe (thus Function weren't
thread-safe until this patch), so simple function copying could
cause memory problems.
